### PR TITLE
Eio bindings Zmq

### PR DESCRIPTION
**Pull request by andersfugmann**
Eio aware bindings for zmq. 

The changes includes factoring out concurrent bindings (i.e. lwt/async/eio) to allow Eio bindings to follow the same interface.
Commits can be read one by one. 

During testing I was puzzled that the Eio tests were significantly slower than the lwt/async tests, however this was due to a fault in the original tests where the delay in sending / receiving was only applied for the first loop iteration. This is also fixed.

### CODE REVIEW

**Comment by andersfugmann** [comment]
I don't want to just add reviewers, so I kindly ask either @Leonidas-from-XIV or @AndreasDahl to give a review on this.

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -42,9 +42,16 @@ jobs:
...
       - name: zmq-async
-        if: ${{ matrix.ocaml-version != 'ocaml-base-compiler.5.0.0~beta1' }}
         run: |
           opam pin add zmq-async.dev . --no-action
           opam install --deps-only --with-doc --with-test zmq-async
           opam exec -- dune build zmq-async
           opam exec -- dune runtest zmq-async
+
+      - name: zmq-eio
+        if: ${{ matrix.ocaml-version == '5.0.0' }}
```
Comment (Leonidas-from-XIV) [inline .github/workflows/workflow.yml:52]: I would maybe set this to `>=` for if the matrix adds 5.1.0.

File: zmq-eio/src/socket.ml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      senders = Queue.create ();
+      receivers = Queue.create ();
+      mutex = Eio.Mutex.create ();
+      condition = Eio.Condition.create ();
+      ready_condition = Eio.Condition.create ();
+      thread = None;
+    }
+  in
+  let thread = Eio.Fiber.fork_promise ~sw (fun () ->
+    Eio.Switch.run (fun sw ->
```
Comment (Leonidas-from-XIV) [inline zmq-eio/src/socket.ml:80]: I don't understand the Eio terms here, what does the `Switch` do?

File: zmq-eio/test/dune [hunk 3]
```diff
@@ -0,0 +1,11 @@
+(executable
+ (name test)
+ (libraries zmq-eio ounit2 eio eio_main))
+
+(rule
+ (alias runtest)
+ (deps
+  (:test test.exe))
```
Comment (Leonidas-from-XIV) [inline zmq-eio/test/dune:8]: `run` implicitly creates a dependency on `test`, so this could be skipped.

File: zmq-eio/test/test.ml [hunk 4]
```diff
@@ -0,0 +1,141 @@
+open OUnit
+
+(** Sleep is horrible slow, or causing strange slowdowns *)
```
Comment (Leonidas-from-XIV) [inline zmq-eio/test/test.ml:3]: This comment is quite unclear. Maybe worth a report upstream?

**Review by Leonidas-from-XIV — COMMENTED** [review]
I am not sure I am the best person to review this, since I don't know very much about Eio but here's at least my observations from the code.

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (andersfugmann) [reply .github/workflows/workflow.yml:52]: Good suggestion. I will update the code.

File: zmq-eio/src/socket.ml (hunk 2 above)
Comment (andersfugmann) [reply zmq-eio/src/socket.ml:80]: Fibers is a lightweight thread. A switch is a collection of fibers and will run until all fibers has ended or cancelled. Daemon fibers will automatically be cancelled once all "regular" fibers has ended. 

The construct here is made to control lifetime of the two fibers started. Once the main loop exits, the signalling daemon fiber will be automatically be cancelled. Creating a promise allows the caller of `close` to wait until the event loop has ended (all handlers completed).

Close is meant to be a nice way to close the socket allowing all readers and writers to end. Its also possible to cancel a switch. If the parent switch (passed as arg this function) is cancelled all child fibers and switches are cancelled. 

I hope this explains the construct of the fork -> switch -> fork.

File: zmq-eio/test/dune (hunk 3 above)
Comment (andersfugmann) [reply zmq-eio/test/dune:8]: I'm not sure I understand. Can the rule be skipped altogether?

File: zmq-eio/test/test.ml (hunk 4 above)
Comment (andersfugmann) [reply zmq-eio/test/test.ml:3]: Well. This was indeed not the case and the comment should be removed. It was before I discovered the bug in the existing lwt/async tests.

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (andersfugmann) [reply .github/workflows/workflow.yml:52]: Ahh. Bit too quick there :-) `>=` will do lexicographical comparison on strings, not test semantic versions. This means that `5.0.10` <= `5.0.9`. I will try to change the test matrix to make it less error prone.

File: zmq-eio/test/test.ml (hunk 4 above)
Comment (andersfugmann) [reply zmq-eio/test/test.ml:3]: I will remove the wrong comment

**Comment by andersfugmann** [comment]
Thanks a lot for your comments @Leonidas-from-XIV - Really appreciated! 

One question about close semantics. In lwt/async, if the socket is closed any deferreds waiting on either `send` or `recv` will never be resolved. 

In `zmq-eio`, `close` will wait for all current sends and receives. Any requests made after the socket has been closed will raise `Closed`. I'm considering changing the semantics to raise `Closed` for all pending requests also when/if the socket is closed. What do you think? What semantics would you expect?

**Review by hongy20 — APPROVED** [review]

**Pull request merged**